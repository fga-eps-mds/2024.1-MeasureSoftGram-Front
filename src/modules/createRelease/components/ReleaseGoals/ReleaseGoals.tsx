import React, { useEffect } from 'react';

import Equalizer from '@modules/createRelease/components/Equalizer';

import { useCreateReleaseContext } from '@modules/createRelease/context/useCreateRelease';
import {
  Box,
  Breadcrumbs,
  Button,
  Checkbox,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  Link,
  Typography
} from '@mui/material';
import * as Styles from './styles';
import useDynamicBalance from './hook/useDynamicBalance';

export default function ReleaseGoals(props: any) {
  const { setActiveStep } = props;
  const { releaseInfoForm, setAllowDynamicBalance } = useCreateReleaseContext();
  const { characteristics, endDate, name, startDate } = releaseInfoForm;
  const { open, allowDynamicBalance, handleChange, handleClose, handleConfirm } = useDynamicBalance();

  useEffect(() => {
    setAllowDynamicBalance(allowDynamicBalance);
  }, [allowDynamicBalance]);

  return (
    <>
      <Styles.Header>
        <h1 style={{ color: '#33568E', fontWeight: '500' }}>Planejamento de Release</h1>
        <h2 style={{ color: '#33568E', fontWeight: '300' }}>Balancear a Meta de Qualidade</h2>
        <Breadcrumbs separator={<Box
          component="span"
          sx={{
            width: 4,
            height: 4,
            borderRadius: '50%',
            bgcolor: 'text.disabled',
          }}
        />} sx={{ fontSize: '14px' }}>

          <Link sx={{
            cursor: 'pointer',
            textDecoration: 'none',
          }} onClick={() => {
            setActiveStep(0);
          }} color="text.secondary">Criar Release</Link>
          <Link onClick={() => {
            setActiveStep(1);
          }} sx={{
            cursor: 'pointer',
            textDecoration: 'none',
          }} color="text.secondary">Definir configuração do modelo</Link>
          <Link sx={{
            cursor: 'pointer',
            textDecoration: 'none',
            fontWeight: '800'
          }} color="text.primary">Balancear Pesos</Link>

        </Breadcrumbs>
        <p>
          Nesta etapa é possível estabelecer metas para as características que serão observadas na release. É
          importante ressaltar que essas características de qualidade estão inter-relacionadas, ou seja, ao modificar a
          meta de uma característica, as demais também serão afetadas.
        </p>
        <p>
          Ao clicar em 'Permitir o balanceamento dinâmico', as alterações feitas em uma característica não afetarão
          automaticamente as demais, permitindo uma flexibilidade maior no planejamento da release.
        </p>
        <Box>
          <Typography color="#FF4646">
            <Checkbox checked={allowDynamicBalance} onChange={handleChange} name="allowDynamicBalanceCheckbox"
              sx={{
                color: '#474747',
                '&.Mui-checked': {
                  color: '#474747',
                },
              }} />
            Permitir o balanceamento dinâmico
          </Typography>
        </Box>
        <Dialog open={open} onClose={handleClose}>
          <DialogTitle>Permitir o balanceamento dinâmico</DialogTitle>
          <DialogContent>
            <Typography color="#FF4646">
              O balanceamento das metas funciona com base em uma matriz de relacionamento entre as características de
              qualidade{' '}
              <Link
                href="https://www.sce.carleton.ca/faculty/wainer/papers/art%253A10.1007%252Fs13198-016-0546-8.pdf"
                underline="hover"
                target="_blank"
              >
                (link)
              </Link>
              . Ao permitir o balanceamento dinâmico, o sistema faz com que essas relações sejam ignoradas, dessa forma,
              alguns objetivos definidos podem ser inalcançáveis. Deseja continuar mesmo assim?
            </Typography>
          </DialogContent>
          <DialogActions>
            <Button onClick={handleClose}>Cancelar</Button>
            <Button onClick={handleConfirm} color="primary" name="confirmButton">
              Confirmar
            </Button>
          </DialogActions>
        </Dialog>
      </Styles.Header>

      <Styles.Body>
        <div>
          <strong>{name}</strong>
          <p>
            {startDate} - {endDate}
          </p>
        </div>

        <Equalizer selectedCharacteristics={characteristics} allowDynamicBalance={allowDynamicBalance} />
      </Styles.Body>
    </>
  );
}
