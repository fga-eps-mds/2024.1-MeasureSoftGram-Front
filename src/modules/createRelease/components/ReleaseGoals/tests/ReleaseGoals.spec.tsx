import '@testing-library/jest-dom';

import React from 'react';
import { fireEvent, render, renderHook } from '@testing-library/react';

import { act } from 'react-dom/test-utils';
import ReleaseGoals from '../ReleaseGoals';
import useDynamicBalance from '../hook/useDynamicBalance';

jest.mock('@modules/createRelease/context/useCreateRelease', () => ({
  useCreateReleaseContext: () => ({
    releaseInfoForm: {
      characteristics: ['reliability', 'maintainability'],
      endDate: '2022-09-25',
      description: 'asdasd',
      name: 'asdasd',
      startDate: '2022-09-18'
    },
    handleChangeForm: jest.fn(),
    setAllowDynamicBalance: jest.fn(),
  })
}));

describe('<ReleaseGoals />', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('Snapshot', () => {
    it('Deve corresponder ao Snapshot', () => {
      const tree = render(<ReleaseGoals />);
      expect(tree).toMatchSnapshot();
    });
  });

  describe('Functions', () => {
    it('Deve chamar as funções de abrir/fechar corretamente', () => {
      const { result } = renderHook(() => useDynamicBalance());

      act(() => result.current.handleChange());
      expect(result.current.open).toBeTruthy();
      act(() => result.current.handleClose());
      expect(result.current.open).toBeFalsy();
      expect(result.current.allowDynamicBalance).toBeFalsy();
      act(() => {
        result.current.handleChange();
        result.current.handleConfirm();
      });
      expect(result.current.allowDynamicBalance).toBeTruthy();
      act(() => result.current.handleChange());
      expect(result.current.allowDynamicBalance).toBeFalsy();
    });
  });
});
