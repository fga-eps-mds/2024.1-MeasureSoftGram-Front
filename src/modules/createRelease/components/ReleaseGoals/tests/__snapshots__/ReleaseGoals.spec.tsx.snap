// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReleaseGoals /> Snapshot Deve corresponder ao Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="styles__Header-sc-ki0fgv-0 dSqkRX"
      >
        <h1
          style="color: rgb(51, 86, 142); font-weight: 500;"
        >
          Planejamento de Release
        </h1>
        <h2
          style="color: rgb(51, 86, 142); font-weight: 300;"
        >
          Balancear a Meta de Qualidade
        </h2>
        <nav
          class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-1gui7g7-MuiTypography-root-MuiBreadcrumbs-root"
        >
          <ol
            class="MuiBreadcrumbs-ol css-4pdmu4-MuiBreadcrumbs-ol"
          >
            <li
              class="MuiBreadcrumbs-li"
            >
              <a
                class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-ae5t40-MuiTypography-root-MuiLink-root"
              >
                Criar Release
              </a>
            </li>
            <li
              aria-hidden="true"
              class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
            >
              <span
                class="MuiBox-root css-1bi6evy"
              />
            </li>
            <li
              class="MuiBreadcrumbs-li"
            >
              <a
                class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-ae5t40-MuiTypography-root-MuiLink-root"
              >
                Definir configuração do modelo
              </a>
            </li>
            <li
              aria-hidden="true"
              class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
            >
              <span
                class="MuiBox-root css-1bi6evy"
              />
            </li>
            <li
              class="MuiBreadcrumbs-li"
            >
              <a
                class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-46jzwd-MuiTypography-root-MuiLink-root"
              >
                Balancear Pesos
              </a>
            </li>
          </ol>
        </nav>
        <p>
          Nesta etapa é possível estabelecer metas para as características que serão observadas na release. É importante ressaltar que essas características de qualidade estão inter-relacionadas, ou seja, ao modificar a meta de uma característica, as demais também serão afetadas.
        </p>
        <p>
          Ao clicar em 'Permitir o balanceamento dinâmico', as alterações feitas em uma característica não afetarão automaticamente as demais, permitindo uma flexibilidade maior no planejamento da release.
        </p>
        <div
          class="MuiBox-root css-0"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-1psv697-MuiTypography-root"
          >
            <span
              class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-12cb8cj-MuiButtonBase-root-MuiCheckbox-root"
            >
              <input
                class="PrivateSwitchBase-input css-1m9pwf3"
                data-indeterminate="false"
                name="allowDynamicBalanceCheckbox"
                type="checkbox"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="CheckBoxOutlineBlankIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </span>
            Permitir o balanceamento dinâmico
          </p>
        </div>
      </div>
      <div
        class="styles__Body-sc-ki0fgv-1 eTbEkS"
      >
        <div>
          <strong>
            asdasd
          </strong>
          <p>
            2022-09-18
             - 
            2022-09-25
          </p>
        </div>
        <div
          class="MuiBox-root css-ckc6o3"
        >
          <div
            class="MuiBox-root css-t0lfuy"
          >
            <div
              class="MuiBox-root css-xi606m"
            >
              <div
                class="MuiBox-root css-vycneo"
              >
                <span
                  class="MuiSlider-root MuiSlider-vertical MuiSlider-colorPrimary MuiSlider-sizeMedium css-y50uh2-MuiSlider-root"
                  data-testid="single-slider"
                >
                  <span
                    class="MuiSlider-rail css-qx5f9e-MuiSlider-rail"
                  />
                  <span
                    class="MuiSlider-track css-ada0pd-MuiSlider-track"
                    style="bottom: 0%; height: 50%;"
                  />
                  <span
                    class="MuiSlider-thumb MuiSlider-thumbSizeMedium MuiSlider-thumbColorPrimary MuiSlider-thumb MuiSlider-thumbSizeMedium MuiSlider-thumbColorPrimary css-y35fuw-MuiSlider-thumb"
                    data-index="0"
                    style="bottom: 50%;"
                  >
                    <input
                      aria-orientation="vertical"
                      aria-valuemax="100"
                      aria-valuemin="0"
                      aria-valuenow="50"
                      data-index="0"
                      max="100"
                      min="0"
                      step="1"
                      style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
                      type="range"
                      value="50"
                    />
                  </span>
                </span>
              </div>
              <p
                class=""
                data-mui-internal-clone-element="true"
              >
                C
                1
              </p>
            </div>
            <div
              class="MuiBox-root css-xi606m"
            >
              <div
                class="MuiBox-root css-vycneo"
              >
                <span
                  class="MuiSlider-root MuiSlider-vertical MuiSlider-colorPrimary MuiSlider-sizeMedium css-y50uh2-MuiSlider-root"
                  data-testid="single-slider"
                >
                  <span
                    class="MuiSlider-rail css-qx5f9e-MuiSlider-rail"
                  />
                  <span
                    class="MuiSlider-track css-ada0pd-MuiSlider-track"
                    style="bottom: 0%; height: 50%;"
                  />
                  <span
                    class="MuiSlider-thumb MuiSlider-thumbSizeMedium MuiSlider-thumbColorPrimary MuiSlider-thumb MuiSlider-thumbSizeMedium MuiSlider-thumbColorPrimary css-y35fuw-MuiSlider-thumb"
                    data-index="0"
                    style="bottom: 50%;"
                  >
                    <input
                      aria-orientation="vertical"
                      aria-valuemax="100"
                      aria-valuemin="0"
                      aria-valuenow="50"
                      data-index="0"
                      max="100"
                      min="0"
                      step="1"
                      style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
                      type="range"
                      value="50"
                    />
                  </span>
                </span>
              </div>
              <p
                class=""
                data-mui-internal-clone-element="true"
              >
                C
                2
              </p>
            </div>
          </div>
          <div
            class="styles__Labels-sc-duigeu-0 hcquOR"
          >
            <p
              data-testid="label"
            >
              C
              1
              : 
              reliability
            </p>
            <p
              data-testid="label"
            >
              C
              2
              : 
              maintainability
            </p>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="styles__Header-sc-ki0fgv-0 dSqkRX"
    >
      <h1
        style="color: rgb(51, 86, 142); font-weight: 500;"
      >
        Planejamento de Release
      </h1>
      <h2
        style="color: rgb(51, 86, 142); font-weight: 300;"
      >
        Balancear a Meta de Qualidade
      </h2>
      <nav
        class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-1gui7g7-MuiTypography-root-MuiBreadcrumbs-root"
      >
        <ol
          class="MuiBreadcrumbs-ol css-4pdmu4-MuiBreadcrumbs-ol"
        >
          <li
            class="MuiBreadcrumbs-li"
          >
            <a
              class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-ae5t40-MuiTypography-root-MuiLink-root"
            >
              Criar Release
            </a>
          </li>
          <li
            aria-hidden="true"
            class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
          >
            <span
              class="MuiBox-root css-1bi6evy"
            />
          </li>
          <li
            class="MuiBreadcrumbs-li"
          >
            <a
              class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-ae5t40-MuiTypography-root-MuiLink-root"
            >
              Definir configuração do modelo
            </a>
          </li>
          <li
            aria-hidden="true"
            class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
          >
            <span
              class="MuiBox-root css-1bi6evy"
            />
          </li>
          <li
            class="MuiBreadcrumbs-li"
          >
            <a
              class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-46jzwd-MuiTypography-root-MuiLink-root"
            >
              Balancear Pesos
            </a>
          </li>
        </ol>
      </nav>
      <p>
        Nesta etapa é possível estabelecer metas para as características que serão observadas na release. É importante ressaltar que essas características de qualidade estão inter-relacionadas, ou seja, ao modificar a meta de uma característica, as demais também serão afetadas.
      </p>
      <p>
        Ao clicar em 'Permitir o balanceamento dinâmico', as alterações feitas em uma característica não afetarão automaticamente as demais, permitindo uma flexibilidade maior no planejamento da release.
      </p>
      <div
        class="MuiBox-root css-0"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-1psv697-MuiTypography-root"
        >
          <span
            class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-12cb8cj-MuiButtonBase-root-MuiCheckbox-root"
          >
            <input
              class="PrivateSwitchBase-input css-1m9pwf3"
              data-indeterminate="false"
              name="allowDynamicBalanceCheckbox"
              type="checkbox"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CheckBoxOutlineBlankIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </span>
          Permitir o balanceamento dinâmico
        </p>
      </div>
    </div>
    <div
      class="styles__Body-sc-ki0fgv-1 eTbEkS"
    >
      <div>
        <strong>
          asdasd
        </strong>
        <p>
          2022-09-18
           - 
          2022-09-25
        </p>
      </div>
      <div
        class="MuiBox-root css-ckc6o3"
      >
        <div
          class="MuiBox-root css-t0lfuy"
        >
          <div
            class="MuiBox-root css-xi606m"
          >
            <div
              class="MuiBox-root css-vycneo"
            >
              <span
                class="MuiSlider-root MuiSlider-vertical MuiSlider-colorPrimary MuiSlider-sizeMedium css-y50uh2-MuiSlider-root"
                data-testid="single-slider"
              >
                <span
                  class="MuiSlider-rail css-qx5f9e-MuiSlider-rail"
                />
                <span
                  class="MuiSlider-track css-ada0pd-MuiSlider-track"
                  style="bottom: 0%; height: 50%;"
                />
                <span
                  class="MuiSlider-thumb MuiSlider-thumbSizeMedium MuiSlider-thumbColorPrimary MuiSlider-thumb MuiSlider-thumbSizeMedium MuiSlider-thumbColorPrimary css-y35fuw-MuiSlider-thumb"
                  data-index="0"
                  style="bottom: 50%;"
                >
                  <input
                    aria-orientation="vertical"
                    aria-valuemax="100"
                    aria-valuemin="0"
                    aria-valuenow="50"
                    data-index="0"
                    max="100"
                    min="0"
                    step="1"
                    style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
                    type="range"
                    value="50"
                  />
                </span>
              </span>
            </div>
            <p
              class=""
              data-mui-internal-clone-element="true"
            >
              C
              1
            </p>
          </div>
          <div
            class="MuiBox-root css-xi606m"
          >
            <div
              class="MuiBox-root css-vycneo"
            >
              <span
                class="MuiSlider-root MuiSlider-vertical MuiSlider-colorPrimary MuiSlider-sizeMedium css-y50uh2-MuiSlider-root"
                data-testid="single-slider"
              >
                <span
                  class="MuiSlider-rail css-qx5f9e-MuiSlider-rail"
                />
                <span
                  class="MuiSlider-track css-ada0pd-MuiSlider-track"
                  style="bottom: 0%; height: 50%;"
                />
                <span
                  class="MuiSlider-thumb MuiSlider-thumbSizeMedium MuiSlider-thumbColorPrimary MuiSlider-thumb MuiSlider-thumbSizeMedium MuiSlider-thumbColorPrimary css-y35fuw-MuiSlider-thumb"
                  data-index="0"
                  style="bottom: 50%;"
                >
                  <input
                    aria-orientation="vertical"
                    aria-valuemax="100"
                    aria-valuemin="0"
                    aria-valuenow="50"
                    data-index="0"
                    max="100"
                    min="0"
                    step="1"
                    style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
                    type="range"
                    value="50"
                  />
                </span>
              </span>
            </div>
            <p
              class=""
              data-mui-internal-clone-element="true"
            >
              C
              2
            </p>
          </div>
        </div>
        <div
          class="styles__Labels-sc-duigeu-0 hcquOR"
        >
          <p
            data-testid="label"
          >
            C
            1
            : 
            reliability
          </p>
          <p
            data-testid="label"
          >
            C
            2
            : 
            maintainability
          </p>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
